### PR TITLE
[AVRO-3507] Implement Single Object reader

### PR DESCRIPTION
**Pull request by jklamer**
Single object encoding readers. Both generic interfacing only with value, and Specific that takes advantage of traits to deserialize directly to the user's type. 

### Jira

https://issues.apache.org/jira/browse/AVRO-3507

### Tests

- [x] Unit tests
- [x] Interop message test

### Documentation
https://issues.apache.org/jira/browse/AVRO-3513

### CODE REVIEW

File: lang/rust/avro/examples/test_interop_single_object_encoding.rs [hunk 1]
```diff
@@ -58,3 +63,15 @@ fn main() {
         .expect("Should encode");
     assert_eq!(file_message, generated_encoding)
 }
+
+fn test_read() {
+    let file_message = std::fs::read(format!("{}/test_message.bin", RESOURCES_FOLDER))
```
Comment (martin-g) [inline lang/rust/avro/examples/test_interop_single_object_encoding.rs:68]: Is this `read` really needed ?

File: lang/rust/avro/src/error.rs [hunk 2]
```diff
@@ -337,6 +337,9 @@ pub enum Error {
     #[error("wrong magic in header")]
     HeaderMagic,
 
+    #[error("Message Header mismatch. Expected: {0:?}. Actual: {1:?}")]
+    MessageHeaderMismatch([u8; 10], [u8; 10]),
```
Comment (martin-g) [inline lang/rust/avro/src/error.rs:341]: SingleObjectHeaderMismatch ?!

File: lang/rust/avro/src/reader.rs [hunk 3]
```diff
@@ -569,4 +663,158 @@ mod tests {
...
+                        ("a", Value::Long(i)) => a = Some(i),
+                        ("b", Value::Double(d)) => b = Some(d),
+                        ("c", Value::Array(v)) => {
+                            for inner_val in v {
+                                if let Value::String(s) = inner_val {
+                                    c.push(s);
+                                }
+                            }
+                        }
+                        _ => panic!("Unexpected pair"),
```
Comment (martin-g) [inline lang/rust/avro/src/reader.rs:720]:
```suggestion
                        (key, value) => panic!("Unexpected pair: {:?} -> {:?}", key, value),
```

File: lang/rust/avro/src/reader.rs [hunk 4]
```diff
@@ -569,4 +663,158 @@ mod tests {
...
+                    }
+                }
+                TestSingleObjectReader {
+                    a: a.unwrap(),
+                    b: b.unwrap(),
+                    c,
+                }
+            } else {
+                panic!("Should be record value")
```
Comment (martin-g) [inline lang/rust/avro/src/reader.rs:729]:
```suggestion
                panic!("Should be record value: {:?}", obj)
```

File: lang/rust/avro/src/reader.rs [hunk 5]
```diff
@@ -569,4 +663,158 @@ mod tests {
...
+                (
+                    "c".into(),
+                    Value::Array(obj.c.into_iter().map(|s| s.into()).collect()),
+                ),
+            ])
+        }
+    }
+
+    #[test]
+    fn test_single_object_reader() {
```
Comment (martin-g) [inline lang/rust/avro/src/reader.rs:748]:
```suggestion
    fn test_avro_3507_single_object_reader() {
```

File: lang/rust/avro/src/reader.rs [hunk 6]
```diff
@@ -569,4 +663,158 @@ mod tests {
...
+            .expect("Schema should resolve");
+        let val = generic_reader
+            .read_value(&mut to_read)
+            .expect("Should read");
+        let expected_value: Value = obj.into();
+        assert_eq!(expected_value, val);
+    }
+
+    #[test]
+    fn test_reader_parity() {
```
Comment (martin-g) [inline lang/rust/avro/src/reader.rs:778]:
```suggestion
    fn test_avro_3507_reader_parity() {
```

File: lang/rust/avro/examples/test_interop_single_object_encoding.rs (hunk 1 above)
Comment (jklamer) [reply lang/rust/avro/examples/test_interop_single_object_encoding.rs:68]: I can reuse. I'll just need to clone because `test_read` needs the owned value

Comment (jklamer) [reply lang/rust/avro/examples/test_interop_single_object_encoding.rs:68]: just kidding, I can use a reference for test write

Comment (martin-g) [reply lang/rust/avro/examples/test_interop_single_object_encoding.rs]: I didn't understand the intention last evening! All is fine!

**Comment by martin-g** [comment]
Thank you, @jklamer !

**Pull request merged**